architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,495 @@
+
+We should conclude that the resource provider protocol must be supported long term, and also that a hybrid approach must be supported. eg: It must be viable for a single resource provider to support both the imperative approach for some types and the declarative approach for others. This is essential to provide continuity for our own services, and also to ensure that users can opt-in to more power when the tradeoff is right.
+
+It's also helpful to think of declarative resource provider functionality as an a-la-carte menu rather than a binary decision. If we can provide support for declarative validation, declarative versioning, and declarative async processing via recipes, users can choose which of those features to apply and where to fall back to code.
+
+The focus of this design document will be on the new declarative functionality because it's new. Keep in mind that the existing resource provider support will remain in place. We're not talking much about it because it already works.
+
+### High Level Design
+
+User-defined types functionality can be broken down into a few different dataflows and systems.
+
+- **Registration and management of user-defined types**: This is a new API that provides CRUDL behavior for the concept of *resource type*. This applies to all resource types regardless of implementation.
+- **Request path for resources implemented declaratively**: This includes routing, validation, storage, and versioning. We typically refer to this today in Radius as the *frontend*.
+- **Processing for resources implemented declaratively**: This includes Recipes or any other asynchronous processing that occurs as a result of a `PUT` or `DELETE` operation. We typically refer to this today in Radius as the *async-worker*.
+- **Tooling that interacts with the Radius API**: This includes the `rad` CLI, Bicep, and Dashboard. When the set of resource-types in Radius is dynamic, then the tools need to be dynamic as well. 
+- **Tooling to author user-defined type schemas**: This includes any tooling we produce to validate a schema, or to transform a schema into a Bicep extension. 
+
+We can also break these systems down by their components in Radius. 
+- Registration and management of user-defined types is part of UCP. Since UCP is the *front-controller* for Radius and is responsible for routing, it makes logical sense that UCP would implement the management of user-defined types.
+- Request path for declarative resources is either part of UCP or a new component. 
+	- UCP implements routing, and will validate that the incoming request is destined for a valid resource type. 
+	- The other *frontend* concerns (validation, storage, versioning) could either be implemented inside UCP or another new component. 
+- Asynchronous processing using recipes is either part of UCP or a new component.
```

Comment:
**Recipes** is just a part of the async processes we have, right? Why do we say **using recipes** here? Because regular **createOrUpdate** or **delete** will be the main operations here. Within the **createOrUpdate** the recipes might be called.

Reply:
This section is about which components host which functionality. I think it's implied that the async working covers both `PUT` and `DELETE` operations.